### PR TITLE
Add endpoint metadata to developer exception page

### LINKED ISSUE #48208: Add endpoint metadata to developer exception page
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Today the developer exception page includes some information about the endpoint, such as its name, route and order.

It's useful to know the metadata on the endpoint. Issues such as https://github.com/dotnet/aspnetcore/issues/39792 are adding useful string representations of metadata.

### Describe the solution you'd like

The `Route` tab on the developer exception page includes a list of the endpoint's metadata. On the server, it would be gathered by simply calling `ToString` on the contents on the metadata collection.

This information could be collapsed initially on the webpage, and clicking a button expands the metadata list. Note that metadata could be very log (e.g. a generated type with a long name that doesn't have a useful ToString) so there should be a maximum length displayable that is truncated by an ellipsis.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/dotnet/aspnetcore/issues/48208

image

### CODE REVIEW

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml [hunk 1]
```diff
@@ -21,312 +21,341 @@
...
+                        <th>@Resources.ErrorPageHtml_MetadataTypeColumn</th>
+                        <th>@Resources.ErrorPageHtml_MetadataDetail</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    @foreach (var metadata in Model.Endpoint.Metadata)
+                    {
+                        <tr>
+                            <td>@metadata.GetType().Name</td>
+                            <td>@(metadata.ToString())</td>
```
Comment (JamesNK) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:343]: Does Razor escape HTML by default? I want to make sure that a ToString result that returns HTML doesn't accidently break the page.

Comment (JamesNK) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:343]: What happens if metadata has new lines? They should be kept when displayed on the web page.

Comment (JamesNK) [reply src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:343]: Also, what happens if metadata has new lines? They should be kept when displayed on the web page.

Comment (JamesNK) [reply src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:343]: I think `HtmlEncodeAndReplaceLineBreaks` (which the title above uses) should be called here.

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml [hunk 2]
```diff
@@ -21,312 +21,341 @@
...
-    </body>
+                    }
+                </tbody>
+            </table>
+        }
+        else
+        {
+            <p>@Resources.ErrorPageHtml_NoRouteValues</p>
+        }
+        <h2>@Resources.ErrorPageHtml_Metadata</h2>
```
Comment (JamesNK) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:328]: The endpoint metadata section is very related to the endpoint section. Either it should be:
- nested under the endpoint section, e.g. have an h3 or h4 title.
- or the metadata section should be next to endpoint and the title should be "Endpoint metadata`.

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml [hunk 3]
```diff
@@ -21,312 +21,341 @@
...
+        @if (Model.Endpoint?.Metadata != null)
```
Comment (JamesNK) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:329]: Also check that the collection isn't empty, e.g. `@if (Model.Endpoint?.Metadata.Count > 0) { ... }`

**Comment by JamesNK** [comment]
Overall, looks good. Just a few changes needed.

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml [hunk 4]
```diff
@@ -21,312 +21,341 @@
...
+                    <tr>
+                        <th>@Resources.ErrorPageHtml_MetadataTypeColumn</th>
+                        <th>@Resources.ErrorPageHtml_MetadataDetail</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    @foreach (var metadata in Model.Endpoint.Metadata)
+                    {
+                        <tr>
+                            <td>@metadata.GetType().Name</td>
```
Comment (JamesNK) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:342]: Could you add the full type name as a tooltip.

Something like:
```suggestion
                            <td><span title="@metadata.GetType().FullName">@metadata.GetType().Name<span></td>
```
(untested)

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml [hunk 5]
```diff
@@ -21,312 +21,342 @@
...
+                    </tr>
+                </thead>
+                <tbody>
+                    @foreach (var metadata in Model.Endpoint.Metadata)
+                    {
                         <tr>
-                            <th>@Resources.ErrorPageHtml_VariableColumn</th>
-                            <th>@Resources.ErrorPageHtml_ValueColumn</th>
+                            <td><span title="@metadata.GetType().FullName">@metadata.GetType().Name</span></td>
+                            <td>@HtmlEncodeAndReplaceLineBreaks(metadata.ToString())</td>
```
Comment (Kahbazi) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/Views/ErrorPage.cshtml:318]: I include both with `HtmlEncodeAndReplaceLineBreaks` and without in the description. I suggest to keep the one without. The razor would encode the html and omit the new lines.

**Comment by danmoseley** [comment]
Is next action here on us -- @JamesNK are you the right person to do the re-review?

**Review by JamesNK — APPROVED** [review]

**Comment by JamesNK** [comment]
Apologies for the delay in reviewing again. Looks good.

**Pull request merged**